File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/core/Driver.java
/*
 * Copyright 2021 Automate The Planet Ltd.
 * Author: Anton Angelov
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package strategyadvanced.core;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.WebDriverWait;

public class Driver {
    private static WebDriverWait browserWait;
    private static WebDriver browser;

    public static WebDriver getBrowser() {
        if (browser == null) {
            throw new NullPointerException("The WebDriver browser instance was not initialized. You should first call the start() method.");
        }
        return browser;
    }

    public static void setBrowser(WebDriver browser) {
        Driver.browser = browser;
    }

    public static WebDriverWait getBrowserWait() {
        if (browserWait == null || browser == null) {
            throw new NullPointerException("The WebDriver browser wait instance was not initialized. You should first call the start() method.");
        }
        return browserWait;
    }

    public static void setBrowserWait(WebDriverWait browserWait) {
        Driver.browserWait = browserWait;
    }

    public static void startBrowser(BrowserType browserType, int defaultTimeout) {

        switch (browserType) {
            case FIREFOX:
                WebDriverManager.firefoxdriver().setup();
                setBrowser(new FirefoxDriver());
                break;
            case CHROME:
                //setBrowser(new ChromeDriver());
                break;
            case EDGE:
                //setBrowser(new EdgeDriver());
                break;
            default:
                break;
        }

        setBrowserWait(new WebDriverWait(getBrowser(), defaultTimeout));
    }

    public static void startBrowser(BrowserType browserType) {
        startBrowser(browserType, 30);
    }

    public static void startBrowser() {
        startBrowser(BrowserType.FIREFOX);
    }

    public static void stopBrowser() {
        getBrowser().quit();
        setBrowser(null);
        setBrowserWait(null);
    }

    public static void waitForAjax() {
        var javascriptExecutor = (JavascriptExecutor)browser;
        browserWait.until(d -> javascriptExecutor.executeScript("return window.jQuery != undefined && jQuery.active == 0"));
    }

    public static void waitUntilPageLoadsCompletely() {
        var javascriptExecutor = (JavascriptExecutor)browser;
        browserWait.until(d -> javascriptExecutor.executeScript("return document.readyState").toString().equals("complete"));
    }
}

File: dotnet/Conferences/PerfectSystemTestsDesign/Core/Driver.cs
﻿// <copyright file="Driver.cs" company="Automate The Planet Ltd.">
// Copyright 2016 Automate The Planet Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
// <author>Anton Angelov</author>
// <site>http://automatetheplanet.com/</site>
using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Support.UI;

namespace PerfectSystemTestsDesign.Core
{
    public static class Driver
    {
        private static WebDriverWait browserWait;

        private static IWebDriver browser;

        public static IWebDriver Browser
        {
            get
            {
                if (browser == null)
                {
                    throw new NullReferenceException("The WebDriver browser instance was not initialized. You should first call the method Start.");
                }
                return browser;
            }
            private set
            {
                browser = value;
            }
        }

        public static WebDriverWait BrowserWait
        {
            get
            {
                if (browserWait == null || browser == null)
                {
                    throw new NullReferenceException("The WebDriver browser wait instance was not initialized. You should first call the method Start.");
                }
                return browserWait;
            }
            private set
            {
                browserWait = value;
            }
        }

        public static void StartBrowser(PerfectSystemTestsDesign.Core.BrowserTypes browserType = PerfectSystemTestsDesign.Core.BrowserTypes.Firefox, int defaultTimeOut = 30)
        {
            switch (browserType)
            {
                case PerfectSystemTestsDesign.Core.BrowserTypes.Firefox:
                    Driver.Browser = new FirefoxDriver();
                    break;
                case PerfectSystemTestsDesign.Core.BrowserTypes.InternetExplorer:
                    break;
                case PerfectSystemTestsDesign.Core.BrowserTypes.Chrome:
                    Driver.Browser = new ChromeDriver();
                    break;
                default:
                    break;
            }
            BrowserWait = new WebDriverWait(Driver.Browser, TimeSpan.FromSeconds(defaultTimeOut));
        }

        public static void StopBrowser()
        {
            Browser.Quit();
            Browser = null;
            BrowserWait = null;
        }
    }
}